Add input range preconditions in the default mathjs interface.

elmisback opened the pull request:
This PR modifies `src/web/demo.js`, `src/web/demo.rkt`, and `src/web/main.css` to allow users to specify input range preconditions in the default mathjs interface.

Ranges appear for each parsed input variable and are specified through pairs of input boxes. If users enter incorrect values in the ranges (non-numbers, numbers outside the ranges, out-of-order ranges), appropriate error messages are displayed. To avoid showing errors while users are working, no errors are displayed for legitimate partial ranges, but users must specify both sides of each range in order to submit the form. Unfilled ranges are highlighted with light blue. Single-point ranges show a yellow-highlighted warning message, but still allow form submission. Once input, values for the range of a particular variable name are remembered on the window object and can be reused if the user happens to change the expression. Submitting the form turns the ranges into a valid precondition.

The FPCore input interface is still accessible via the extra-links and loads on the same page with the expression and precondition from the mathjs interface. (Internally, the setup_state() function is still used for this.)

A few other interface changes are included.
* Submission using the "Enter" key was removed since it no longer makes sense. 
* The "extra options" inputs were removed, since users who would like to use a more advanced interface can just use the FPCore input box or the command line tool.
* Parsing of input to the mathjs input was debounced with a 400ms delay to avoid showing errors and partial ranges as the user is typing.

pavpanchekha commented:
I looked over the code. My main comment is that right now the code converts the input ranges into a mathjs precondition, which we then parse and munge into an FPCore precondition. Why not cut out the middle step and go directly to an FPCore precondition?

elmisback commented:
Good point! It makes sense to cut out the middle step if there's not going to be a mathjs precondition input anyway. I'll finish making the other changes and then add direct conversion to FPCore as a separate commit on this PR.

elmisback commented:
Updates:
* Removed "back to MathJS" button from FPCore interface; user can use refresh to return.
* Switched to "orange" for warnings.
* "Convert to FPCore input" link renamed and moved to the bottom right of the form; I think this position is more reasonable than the top of the page because it acts like a form submission that includes the preconditions and most users don't need to think about it when first entering values.
* Errors and warnings (including formula parse errors) are grouped so the input boxes don't jump around, and appear right before the Improve with Herbie button to make sure it's clear that addressing them will unlock the button.
* As Pavel suggested, the precondition inputs use a table to keep the input boxes lined up.

elmisback commented:
Tested on Chrome 99, Firefox 98, and Safari 15.3

pavpanchekha commented:
I think this is good to merge! Go ahead as soon as you're ready.

elmisback commented:
I double-checked everything, merging now.
